[13.0][MIG] l10n_es_aeat

MiquelRForgeFlow opened the pull request:
Standard migration.

Commit:
[ADD] l10n_es_aeat, l10n_es_aeat_347, l10n_es_aeat_349: Adds a generic module for aeat models, 347 module was portedto v6.0 and adds new module to print AEAT model 349.

Commit:
[IMP+FIX] l10n_es_aeat, l10n_es_aeat_mod347, l10n_es_aeat_mod349: Pequeñas mejoras generales (vistas, traducciones, código) y corrección pequeño bug por un olvido en la adaptación de la v5 a la v6

albertca committed:
[FIX] l10n_es_aeat: Fix _format_number function so number '22027.0' is formatted correctly. With this number 'dec_part' would become 100 (instead of 0) and fail in the final assertion. Instead of providing a workaround for this case, I simply replace the formatting with '%0*.*f' which seems easier specially because it already does the appropriate rounding.

Commit:
[REF] Cambio de licencia de GPL3 a AGPL en los módulos donde sólo aparecía Pexego en la licencia, quedan los compartidos.

Commit:
[FIX] l10n_es_aeat_mod347,l10n_es_aeat: Añadimos los últimos cambios en la 6.0 de estos módulos

Commit:
Commit inicial de una rama para la versión 7.0. Todos los módulos fueron marcados como no instalables mientras no se compruebe que funcionan o migren, para poder usarlos o probarlos se debería poner el atributo installable de los ficheros __openerp__.py de cada módulo a True

Commit:
Launchpad automatic translations update.

acysos committed:
[ADD] l10n_es_aeat: migración a OpenERP 7.0

Commit:
Launchpad automatic translations update.

Commit:
[FIX] l10n_es_aeat: fix exception call (except_orm instead of orm_exception)

acysos committed:
[FIX] l10n_es_aeat: bug atthach_id

Commit:
[FIX] msgstr in .pot files should be empty

Commit:
Launchpad automatic translations update.

Commit:
[IMP] l10n_es_aeat: Icono para el módulo.
- Algo de PEP8 y refactorización de código

[MIG] l10n_es_aeat_mod303: Migración a v7 del módulo con las siguientes mejoras:
- Convenciones de la comunidad.
- PEP8 en la mayoría de código.
- Icono para el módulo.
- Adaptación de vistas a las posibilidades de la v7.
- Restringidas las cuentas bancarias a las de la compañía.

Commit:
Launchpad automatic translations update.

Commit:
[IMP] l10n_es_aeat: Mejorado mecanismo de herencia para reaprovechar más código.
- Añadido wizard para ver el resultado de la exportación sin necesidad de tener el módulo document.
- Puesta transición de estados sin workflow.
- Pequeños ajustes de interfaz.
[IMP] l10n_es_aeat_mod303: Adaptado al nuevo mecanismo. Se elimina workflow anterior mediante un script de migración.

Commit:
Launchpad automatic translations update.

acysos committed:
[FIX] l10n_es_aeat: compatibilidad de la función export_to_boe para el modelo 340

Commit:
[FIX] l10n_es_aeat: Pequeños cambios para acomodar la funcionalidad del 347 y el 340.

Commit:
Launchpad automatic translations update.

Commit:
[IMP] l10n_es_aeat:

* Eliminación de la mayoría de warnings que da runbot.
* Usar librería unidecode en lugar de reemplazos manuales para las tildes y caracteres no soportados.

Commit:
Launchpad automatic translations update.

pedrobaeza committed:
[MIG] Movidos todos los módulos a __unported__
[MIG] Eliminados módulos no migrados a v7

pedrobaeza committed:
[FIX] l10n_es_aeat: New PEP8 checks
- Mover import de unidecode para evitar error

Commit:
[IMP] l10n_es_aeat:

* unificar campos en el módulo base + migrar a la nueva api.
* README.rst mejorado
* mejora de código
* usabilidad del atributo _aeat_number
* secuencia automatica
* control al instanciar los modelos
* correcciones necesarias para el correcto funcionamiento de los módulos

pedrobaeza committed:
[FIX] l10n_es_aeat_*: PEP8 + Refinamientos varios

pedrobaeza committed:
[IMP+FIX] l10n_es_aeat_*: Cálculo por códigos de impuestos generalizado + tratamiento residentes 216 y cálculo

pedrobaeza committed:
[MIG] l10n_es_aeat_mod303: Modelo 303 de la AEAT totalmente refactorizado

Presentación del Modelo AEAT 303
================================

* Incluye un motor genérico para obtener los totales de códigos de impuestos
  de manera parametrizada y con posibilidad de versionado por fechas
* Incluye un motor de parametrización de las exportaciones a formato BOE
  con posibilidad de versionado por fechas.
* Se incluyen las parametrizaciones de ambos motores para el 303.

pedrobaeza committed:
[FIX] l10n_es_aeat_mod303: Problema de permisos en l10n.es.aeat.tax.line

oca-transbot committed:
OCA Transbot updated translations from Transifex

pedrobaeza committed:
[IMP] l10n_es_aeat_vat_prorrate: Enfoque definitivo

======================
AEAT - Prorrata de IVA
======================

Módulo para gestionar la prorrata del IVA en las declaraciones de la AEAT (por
el momento, en el modelo 303), según el artículo 92 de la Ley 37/1992, de 28 de
diciembre, del Impuesto sobre el Valor Añadido.

Uso
===

Realizando el modelo 303, se puede seleccionar si se desea aplicar prorrata de
IVA.

Prorrata general
----------------

En el caso de seleccionar "Prorrata general", aparece un nuevo campo llamado
"Porcentaje de prorrata de IVA" que puede ser rellenado a mano.

Además, cuando se realice el asiento de regularización, la parte proporcional
de la cuota no deducida se compensará como gasto.

Se ha incluido un calculador del porcentaje de prorrata, utilizable en 2 casos:

* Al comienzo de un nuevo ejercicio, para determinar el porcentaje provisional
  a aplicar hasta el último periodo de ese ejercicio.
* En la última declaración del ejercicio, para determinar el porcentaje
  definitivo de prorrata, y sobre el que se compensará la diferencia.

Problemas conocidos / Hoja de ruta
==================================

* La prorrata especial de IVA no está contemplada aún.

pedrobaeza committed:
[IMP] l10n_es_aeat/l10n_es_aeat_mod303: Crear asiento de regularización cuando se hace la declaración

* Generador del asiento de regularización con cargo a un proveedor "Agencia
  Estatal de Administración Tributaria" creado al efecto.

Configuración
=============

Para activar la creación del asiento de regularización en un modelo, hay que
poner en el modelo correspondiente el campo allow_posting a True, y establecer
en la configuración de impuestos los conceptos que se regularizarán con el
flag "to_regularize".

oca-transbot committed:
OCA Transbot updated translations from Transifex

pedrobaeza committed:
[FIX] l10n_es_aeat_mod111: Eliminar periodo específico y corregir exportación

oca-transbot committed:
OCA Transbot updated translations from Transifex

pedrobaeza committed:
[IMP] l10n_es_aeat_*: Cálculo de importes de impuestos por cadena en lugar de recordset + gestionar grupos de compañías

pedrobaeza committed:
[FIX] l10n_es_aeat: Remove tax code search limit

pedrobaeza committed:
[FIX] l10n_es_aeat: _formatNumber con 0 para positivos en lugar de espacio. Fixes #228

oca-transbot committed:
OCA Transbot updated translations from Transifex

jcarlosmontoya committed:
[FIX]l10n_es_aeat_export_to_boe_button

PR related with: https://github.com/OCA/l10n-spain/issues/236

oca-transbot committed:
OCA Transbot updated translations from Transifex

pedrobaeza committed:
[FIX] l10n_es_aeat: Default again to ' ' in positive numbers and allow other values in export config

oca-transbot committed:
OCA Transbot updated translations from Transifex

Commit:
[8.0][FIX] l10n_es_aeat: Corrige creación de apuntes con importe en debe y haber a la vez.

pedrobaeza committed:
[FIX] l10n_es_aeat_*: Mejor herencia de impuestos

El anterior método utilizado, al tener una clase y un many2one genérico,
no permitía poner los datos en múltiples modelos hijos, ya que se repetían
los IDs, y cogía los datos de un modelo en otros.

Con este nuevo sistema, se sigue utilizando una tabla, pero con doble
índice: el modelo y el ID, parecido a un campo reference, pero gestionado
por código propio en el método de cálculo.

El 303 está adaptado a este nuevo sistema, y se incluye un script de
migración que conserve los datos.

pedrobaeza committed:
[IMP] l10n_es_aeat_report: Remove tax line when report is removed

oca-transbot committed:
OCA Transbot updated translations from Transifex

antespi committed:
[FIX] Do not use hyphen in report sequence, only numbers
- Change field name: 'sequence' by 'name'
- Isolate report identifier number calculation in order to be inheritable
- Change methos name: _name_get by _report_identifier_get
- Remove sequence constraints

oca-transbot committed:
OCA Transbot updated translations from Transifex

pedrobaeza committed:
[IMP] l10n_es_aeat: Don't split regularization lines for same account. Closes #305

If some tax lines have to been regularized, and they have the same account, previous
to this commit, they appear in separate lines. Now, all are merged in only one line
(if the balance sign is the same).

oca-transbot committed:
OCA Transbot updated translations from Transifex

pedrobaeza committed:
[FIX] l10n_es_aeat: Sequence number multi-company

oca-transbot committed:
OCA Transbot updated translations from Transifex

Roodin committed:
[FIX].l10n_es_aeat. Correccion de Warning en onchange de period_type

Roodin committed:
l10n_es_aeat,  l10n_es_aeat_mod349. Pequeñas mejoras

Roodin committed:
l10n_es_aeat. Pequeños retoques de código, según correcciones propuestas

pedrobaeza committed:
[FIX] l10n_es_aeat: Admit only valid character set on formatString

oca-transbot committed:
OCA Transbot updated translations from Transifex

Commit:
[IMP] l10n_es_aeat_mod111: Nuevo formato de exportación del modelo 111 para los ejercicios 2016 y siguientes

pedrobaeza committed:
[FIX] l10n_es_aeat: Handle non string variables in _formatString - Fixes #375

cubells committed:
[IMP] Add tax amount to view

oca-transbot committed:
OCA Transbot updated translations from Transifex

cubells committed:
[FIX] statusbar_colors not supported

pedrobaeza committed:
[MIG] l10n_es_aeat: Initial migration

antespi committed:
[MIG] l10n_es_aeat/303: Rest of the migration

oca-transbot committed:
OCA Transbot updated translations from Transifex

pedrobaeza committed:
[FIX] l10n_es_aeat: Bump version

antespi committed:
[IMP] Tomar el último día del ultimo periodo para el asiento de regularización

antespi committed:
[8.0][IMP][l10n_es_aeat_mod347] Improve performance and other things (#317)

angelmoya committed:
[FIX] l10n_es_aeat: Make repeat_expresion work

oca-transbot committed:
OCA Transbot updated translations from Transifex

pedrobaeza committed:
[IMP] l10n_es_aeat: Several warnings and imps

pedrobaeza committed:
[MIG] l10n_es_aeat/mod303: Migration to 10.0

pedrobaeza committed:
[FIX] l10n_es_aeat: Include direct statement account moves with taxes

When you fill a bank or cash statement with taxes for cancelling an expense,
those amounts weren't included on the reports, as the move type is
'liquidity', not 'receivable' or 'payable'. This fix includes also this type
on the search domain.

pedrobaeza committed:
[FIX] l10n_es_aeat: Correct constraint for multi-company

pedrobaeza committed:
[FIX] l10n_es_aeat_mod303: Adapt new BOE 2017 format

pedrobaeza committed:
[IMP] l10n_es_aeat: Avoid rewrites on related fields

ORM launches a rewrite on origin table for related fields without
readonly flag. Preventing this marking it as readonly, we will
gain a bit in performance or avoid write errors.

pedrobaeza committed:
[FIX] l10n_es_aeat: Avoid multi-company error due to depends

Debido a que se utiliza una sola tabla para almacenar las líneas
de los impuestos de todos los modelos, hay una limitación en el ORM de Odoo
cuando se coloca el campo one2many de dichas líneas (tax_line_ids) como
dependencia en la definición del cálculo de un campo (entrada con
@api.depends), que recalcula los campos calculados de todos los modelos con el
mismo ID que el del registro en curso, lo que puede ser un problema en entornos
multi-compañía. Una solución a ello (aunque no evita el recálculo), es poner en
esos campos calculados `compute_sudo=True`.

pedrobaeza committed:
[IMP] l10n_es_aeat: Check previous declaration number on proper types

If the declaration is complementary or substitutive, there should be
a previous declaration number.

Roodin committed:
[FIX]. l10n_es_aeat. Fijado problema con referencai a _model al exportar

pedrobaeza committed:
[FIX] l10n_es_aeat_*: Conditional expression is reversed + 303 with conditional

gurneyalex committed:
[REF] i18n: Updating translation terms from weblate [ci skip]

oca-transbot committed:
OCA Transbot updated translations from Transifex

pedrobaeza committed:
[FIX] l10n_es_aeat: Workaround for odoo/odoo#18438

Due to a bug in ORM that unlinks other tables' records, we have to avoid (0, 0, x) syntax

Reference: https://github.com/odoo/odoo/issues/18438

oca-transbot committed:
OCA Transbot updated translations from Transifex

pedrobaeza committed:
[IMP] l10n_es_aeat: Add method for formatting BOE dates

misern2 committed:
[MIG] l10n_es_aeat_mod115: Migration  to 10.0

pedrobaeza committed:
[IMP] l10n_es_aeat: _register_hook without warning

Calling it with decorator @api.model_cr

oihane committed:
[MIG] l10n_es_aeat: Migration to version 11.0

Commit:
[FIX] l10n_es_aeat: ljust exige byte string de lenght 1

hugosantosred committed:
[FIX] l10n_es_aeat: Bytestring concatenation on file export to boe (Solves compatibility problems with python3) (#1)
* Adds tests for exporting files to boe with an export config (#2)

hugosantosred committed:
[FIX] l10n_es_aeat: Add onchange_chart_template_id to TestL10nEsAeatModBase._chart_of_accounts_create() in order to fix travis tests
* remove fax, field does not exist

JordiBForgeFlow committed:
[FIX] l10n_es_aeat: Fix BOE export (#718)

oca-transbot committed:
OCA Transbot updated translations from Transifex

JordiBForgeFlow committed:
[MIG] l10n_es_aeat_mod349: Migration to 10.0

pedrobaeza committed:
[MIG] l10n_es_aeat_mod349: Migration to 11.0

pedrobaeza committed:
[IMP] l10n_es_aeat_mod303: Improve tests for handling negative price units

pedrobaeza committed:
[FIX+IMP] l10n_es_aeat_mod303: Fix refinements

* No need to change test structure with create
* Optimize code
* Fix tests in l10n_es_vat_prorrate
* Translations
* Other minor things

pedrobaeza committed:
[IMP] l10n_es_aeat_mod303: Date for regularization move on the end of period

JordiBForgeFlow committed:
migración de l10n_es_vat_book a 11.0

pedrobaeza committed:
[IMP] l10n_es_aeat: Add tax_exigibility option to tax mapping

With this, cash basis can be done in an easy way, and it's a base field
in account module.

Abranes committed:
[IMP] l10n_es_aeat: New export_type alphabetic

oca-travis committed:
[UPD] Update l10n_es_aeat.pot

mtbochaca committed:
Translated using Weblate (Catalan)

Currently translated at 80.9% (144 of 178 strings)

Translation: l10n-spain-11.0/l10n-spain-11.0-l10n_es_aeat
Translate-URL: https://translation.odoo-community.org/projects/l10n-spain-11-0/l10n-spain-11-0-l10n_es_aeat/ca/

jvpascual committed:
[11.0] [IMP] l10n_es_aeat: Contact email added. (#866)

oca-travis committed:
[UPD] Update l10n_es_aeat.pot

pedrobaeza committed:
[FIX] l10n_es_aeat: Adapt patch to P3

oca-travis committed:
[UPD] Update l10n_es_aeat.pot

oca-travis committed:
[UPD] Update l10n_es_aeat.pot

pedrobaeza committed:
[FIX] l10n_es_aeat_mod303: Handle negative amounts in purchase refunds

It includes tests for the new cases.

pedrobaeza committed:
[MIG] l10n_es_aeat: Migration to 11.0

- Standard procedure
- Adapt views
- Adapt tests
- README by fragments

pedrobaeza committed:
[MIG] l10n_es_aeat_mod303: Migration to 12.0

* Standard procedure
* Adapted tests
* README by fragments

oca-travis committed:
[UPD] Update l10n_es_aeat.pot

pedrobaeza committed:
[FIX] l10n_es_aeat: Don't put company to AEAT partner

So it works for multi-company environments.

oca-travis committed:
[UPD] Update l10n_es_aeat.pot

pedrobaeza committed:
[IMP] l10n_es_aeat_mod303: Add support for 2018 4T:

* BOE format
* Exonerated from 390

  * Fields
  * Mappings
  * BOE export

* README update
* Little change in AEAT engine
* Tests adapted + expanded

pedrobaeza committed:
[IMP] l10n_es_aeat: Add BOE file comparer

New feature for viewing a BOE exported file with each field properly mapped
to their export config line for showing the values.

oca-travis committed:
[UPD] Update l10n_es_aeat.pot

pedrobaeza committed:
[IMP] l10n_es_aeat: Add active field to export configs

This way, we don't initially see all the subconfigs for each model and we can concentrate
on main export configurations. This is got through `active` field default behaviour.
Archived config are still directly searchable through the subconfig selector in the
configuration line.

All models export configs have been adapted to this new approach.

oca-travis committed:
[UPD] Update l10n_es_aeat.pot

mvrodriguez committed:
Translated using Weblate (Spanish)

Currently translated at 100.0% (194 of 194 strings)

Translation: l10n-spain-12.0/l10n-spain-12.0-l10n_es_aeat
Translate-URL: https://translation.odoo-community.org/projects/l10n-spain-12-0/l10n-spain-12-0-l10n_es_aeat/es/

mvrodriguez committed:
Translated using Weblate (Galician)

Currently translated at 100.0% (194 of 194 strings)

Translation: l10n-spain-12.0/l10n-spain-12.0-l10n_es_aeat
Translate-URL: https://translation.odoo-community.org/projects/l10n-spain-12-0/l10n-spain-12-0-l10n_es_aeat/gl/

pedrobaeza committed:
[FIX+IMP] l10n_es_aeat_mod347: Migration refinements + enhancements

* Add controller for automatic confirmation/rejection
* Improve data model
* Clean code
* Clean UI
* Spanish translation
* Tests
* ...

oca-travis committed:
[UPD] Update l10n_es_aeat.pot

oca-travis committed:
[UPD] Update l10n_es_aeat.pot

angelmoya committed:
[FIX] res_partner_bank.acc_type type of field

oca-travis committed:
[UPD] Update l10n_es_aeat.pot

pedrobaeza committed:
[FIX] l10n_es_aeat: Create sequences when new company is created

When a new company is created, its sequences are not populated, so an Odoo service
restart is needed for being able to use AEAT models. With this, you don't need
this restart.

It includes a test with a hack for populating a model in the test itself for testing
complete flow.

oca-travis committed:
[UPD] Update l10n_es_aeat.pot

pedrobaeza committed:
[FIX] l10n_es_aeat: Depends on l10n_es

This dependency is not strictly needed by the module, but the test base for other
modules needs the module, so let's put it explicitly

pedrobaeza committed:
[IMP] l10n_es_aeat: Move state mapping to base

oca-travis committed:
[UPD] Update l10n_es_aeat.pot

pedrobaeza committed:
[FIX] l10n_es_aeat: Remove " from BOE files

It's not accepted by AEAT.

oca-travis committed:
[UPD] Update l10n_es_aeat.pot

pedrobaeza committed:
[FIX] l10n_es_aeat_mod347: Proper variable assignation

So that inherited models are able to access this value.

oca-travis committed:
[UPD] Update l10n_es_aeat.pot

josepmy committed:
Translated using Weblate (Spanish)

Currently translated at 100.0% (197 of 197 strings)

Translation: l10n-spain-12.0/l10n-spain-12.0-l10n_es_aeat
Translate-URL: https://translation.odoo-community.org/projects/l10n-spain-12-0/l10n-spain-12-0-l10n_es_aeat/es/

pedrobaeza committed:
[IMP] l10n_es_aeat: Add chatter + activities to AEAT models

* Include chatter (mail.thread)
* Include activities (mail.activity.mixin)
* Track AEAT report state change
* Include new admin user in AEAT group by default

oca-travis committed:
[UPD] Update l10n_es_aeat.pot

pedrobaeza committed:
Translated using Weblate (Spanish)

Currently translated at 100.0% (228 of 228 strings)

Translation: l10n-spain-12.0/l10n-spain-12.0-l10n_es_aeat
Translate-URL: https://translation.odoo-community.org/projects/l10n-spain-12-0/l10n-spain-12-0-l10n_es_aeat/es/

carlosdauden committed:
[IMP] l10n_es_aeat: Añadir campo aeat_anonymous_cash_customer y funcion parse_vat_info

oca-travis committed:
[UPD] Update l10n_es_aeat.pot

carlosdauden committed:
[IMP] l10n_es_aeat: Añadir método que devuelve impuestos desde plantillas

oca-travis committed:
[UPD] Update l10n_es_aeat.pot

sergio-teruel committed:
[FIX] l10n_es_aeat: fix report client action launch

CarlesAntoli committed:
Translated using Weblate (Catalan)

Currently translated at 96.1% (222 of 231 strings)

Translation: l10n-spain-12.0/l10n-spain-12.0-l10n_es_aeat
Translate-URL: https://translation.odoo-community.org/projects/l10n-spain-12-0/l10n-spain-12-0-l10n_es_aeat/ca/

acysos committed:
[ADD] l10n_es_aeat: certificates and soap. AEAT Partner check

[ADD] l10n_es_aeat: certificates and soap. AEAT Partner check

[ADD] l10n_es_aeat: certificates and soap. AEAT Partner check

[ADD] l10n_es_aeat: certificates and soap. AEAT Partner check

[ADD] l10n_es_aeat: certificates and soap. AEAT Partner check

oca-travis committed:
[UPD] Update l10n_es_aeat.pot

MiquelRForgeFlow committed:
[UPD] Revert obsolete workaround

It was solved in v9 (see https://github.com/odoo/odoo/pull/19181/commits/9ae78328ec6fa0c1ba5c0a5f7583a565c7fa4e3f).

MiquelRForgeFlow committed:
[UPD] Remove deprecated from_string()

MiquelRForgeFlow committed:
[UPD] Remove deprecated decodestring and encodestring

MiquelRForgeFlow committed:
[IMP] Get taxes from templates instead of code description

oca-travis committed:
[UPD] Update l10n_es_aeat.pot

MiquelRForgeFlow committed:
[IMP] l10n_es_aeat: black, isort

MiquelRForgeFlow committed:
[MIG] l10n_es_aeat: Migration to 13.0

HaraldPanten commented:
Hola a todos y gracias @MiquelRForgeFlow por la iniciativa.

Será éste el PR que sustituya a: https://github.com/OCA/l10n-spain/pull/1189 , finalmente?

No estoy seguro de si lo iba a retomar el equipo de Acysos con un PR suyo que comentaban que es operativo. @acysos @pedrobaeza ¿Podéis arrojar algo de luz?

Saludos!

acysos commented:
@HaraldPanten nosotros lo tenemos hecho, fui a publicarlo ayer cuando me encontre este, pero parece que no lo leyeron, doble trabajo realizado. Yo de momento paro, para no duplicar trabajo, el resto de los módulos AEAT también lo tengo, pero a menos @MiquelRForgeFlow diga que los suba no voy a duplicar más trabajo.

HaraldPanten commented:
Por supuesto @acysos , más que nada es ir todos en la misma dirección y no duplicar el trabajo de nadie, como bien dices. Una vez en este punto ¿qué creéis que sería lo más oportuno de cara a avanzar lo más eficientemente posible?

Gracias y saludos.

MiquelRForgeFlow commented:
> pero parece que no lo leyeron

Lo siento, no lo vi en ningún lugar.

El PR ya está revisable e hice PR para el modelo 111 para que puedan comprobar y revisar bien.

HaraldPanten reviewed (approved):
Testeado funcionalmente y en principio me encaja. 👍

HaraldPanten commented on l10n_es_aeat/models/l10n_es_aeat_report_tax_mapping.py:
```diff
@@ -0,0 +1,194 @@
+            ("date", ">=", date_start),
+            ("date", "<=", date_end),
+            ("exclude_from_invoice_tab", "=", True),
+        ]
```
¿El dominio no debería ser?:

move_line_domain = [
            ("company_id", "child_of", self.company_id.id),
            ("date", ">=", date_start),
            ("date", "<=", date_end),
            ("parent_state", '=', 'posted'),
        ]

Se ha eliminado el:

- ("exclude_from_invoice_tab", "=", True),
Debido a que al realizar facturas/asientos/apuntes manuales este campo solamente pasa a True cuando se trata de líneas con Impuestos aplicados, por el resto de casos es False.

Se ha incluido el:

-  ("parent_state", '=', 'posted')
Debido a que no deberían tenerse en cuenta aquellos apuntes donde el padre (el asiento) no está publicado aún, ya que puede dar a errores/problemas en los cálculos de los modelos.

MiquelRForgeFlow replied:
me parece bien. era algo con lo que dudaba. gracias

HaraldPanten commented on l10n_es_aeat/models/l10n_es_aeat_report_tax_mapping.py:
```diff
@@ -0,0 +1,194 @@
+            [
+                ("id", "in", tax_lines.move_line_ids.ids),
+                ("exclude_from_invoice_tab", "=", True),
+            ],
```
¿El dominio no debería ser?:
[
            ("id", "in", tax_lines.move_line_ids.ids),
            ("parent_state", '=', 'posted'),
 ]

Se ha eliminado el:

- ("exclude_from_invoice_tab", "=", True),
Debido a que al realizar facturas/asientos/apuntes manuales este campo solamente pasa a True cuando se trata de líneas con Impuestos aplicados, por el resto de casos es False.

Se ha incluido el:

- ("parent_state", '=', 'posted')
Debido a que no deberían tenerse en cuenta aquellos apuntes donde el padre (el asiento) no está publicado aún, ya que puede dar a errores/problemas en los cálculos de los modelos.

HaraldPanten reviewed (changes requested):
Hola @MiquelRForgeFlow ,

Hemos estado revisando la base conjuntamente con el https://github.com/OCA/l10n-spain/pull/1307 y hemos detectado ciertos problemas en los filtros de apuntes.

Commit:
Translated using Weblate (Spanish)

Currently translated at 100.0% (249 of 249 strings)

Translation: l10n-spain-12.0/l10n-spain-12.0-l10n_es_aeat
Translate-URL: https://translation.odoo-community.org/projects/l10n-spain-12-0/l10n-spain-12-0-l10n_es_aeat/es/

ValentinVinagre commented:
Hola @MiquelRForgeFlow ,
¿has tenido ocasión de revisar el comentario del https://github.com/OCA/l10n-spain/pull/1307?

MiquelRForgeFlow commented:
> ¿has tenido ocasión de revisar el comentario del #1307?

No, lo siento, ando un poco liado. Aparte de @HaraldPanten, ¿alguien mas ha testeado ese problema?

pedrobaeza commented:
Yo aún no

ValentinVinagre commented:
Hola @MiquelRForgeFlow @pedrobaeza ,
nosotros lo probamos y propusimos una posible solución, queda por ver si es viable o es mejor utilizar otro método. En el https://github.com/OCA/l10n-spain/pull/1307

Un saludo,

acysos commented:
Yo estoy testandolo con el resto de modelos AEAT, comprobar que no falla con ninguno y lo apruebo y mando el resto de PR.

HaraldPanten commented:
@acysos Vosotros cómo habéis solventado el comentario de https://github.com/OCA/l10n-spain/pull/1307 ?

¿Es mejor cambiar la `def calculate(self)` en el base l10n_es_aeat? O lo habéis hecho de otro modo?

Saludos,

pedrobaeza committed:
[IMP+FIX] l10n_es_aeat: Move get taxes method to company + cleaning in tests

This way, we can reuse it in other modules that don't inherit AEAT report like SII.

We also clean a non needed reading of the tax description field.

Tests were still using description method for populating data, so we switch to new
system and perform needed changes in dependent tests.

s

ValentinVinagre commented:
Hola @MiquelRForgeFlow ,
¿Puedes resolver los conflictos?
No se si has podido revisar lo que hemos comentado del modelo 111 https://github.com/OCA/l10n-spain/pull/1307

Un Saludo,

oca-travis committed:
[UPD] Update l10n_es_aeat.pot

MiquelRForgeFlow commented:
> ¿Puedes resolver los conflictos?

Hecho

> No se si has podido revisar lo que hemos comentado del modelo 111 #1307

Aun no 😓

ValentinVinagre commented on l10n_es_aeat/tests/test_l10n_es_aeat_mod_base.py:
```diff
@@ -0,0 +1,309 @@
+            data["invoice_line_ids"].append((0, 0, line_data))
+        if extra_vals:
+            data.update(extra_vals)
+        inv = cls.env["account.move"].sudo(cls.billing_user).create(data)
```
El sudo está deprecated, debería ser with_user.

MiquelRForgeFlow replied:
ya, puede que me deje algunos

ValentinVinagre commented on l10n_es_aeat/tests/test_l10n_es_aeat_mod_base.py:
```diff
@@ -0,0 +1,309 @@
+            data["invoice_line_ids"].append((0, 0, line_data))
+        if extra_vals:
+            data.update(extra_vals)
+        inv = cls.env["account.move"].sudo(cls.billing_user).create(data)
```
El sudo está deprecated, debería ser with_user.

ValentinVinagre commented on l10n_es_aeat/tests/test_l10n_es_aeat_mod_base.py:
```diff
@@ -0,0 +1,309 @@
+        return inv
+
+    @classmethod
+    def _invoice_refund(cls, invoice, dt):
```
En la versión 13 no existe la función refund(), yo he realizado la función de la siguiente manera:

```
@classmethod
    def _invoice_refund(cls, invoice, dt):
        _logger.debug("Refund {} invoice: date = {}".format(invoice.type, dt))
        default_values_list = [{
            'date': dt,
            'invoice_date': dt,
            'journal_id': cls.journal_misc.id or invoice.journal_id.id,
            'invoice_payment_term_id': None,
        }]
        inv = invoice.with_user(cls.billing_user)._reverse_moves(
            default_values_list
        )
        inv.post()
        if cls.debug:
            cls._print_move_lines(inv.line_ids)
        return inv
```

MiquelRForgeFlow replied:
es cierto. no entiendo entonces porque ni runbot ni travis petan 🤔

ValentinVinagre reviewed (changes requested):
Hola @MiquelRForgeFlow ,
¿Puedes realizar estos cambios? Estamos con el modelo 303 y con estos arreglo si no voy mal se arreglan bastantes casos.

Gracias,

ValentinVinagre commented:
Hola @pedrobaeza @MiquelRForgeFlow @HaraldPanten ,
Debido al cambio de impuestos que ha realizado Odoo, me he encontrado mientras trabaja sobre el https://github.com/OCA/l10n-spain/pull/1328 que el informe de impuestos que está implementado en la aeat no contempla ciertas casuísticas. Por ej.
Factura:
Linea 1: importe 100.00 €, IVA 21% Adquisición de servicios intracomunitarios

Apuntes de tasas:
100.00  of tax  472000 Hacienda Pública. IVA soportado  
-100.00 of tax  477000 Hacienda Pública. IVA repercutido  

Problema:
Dado que el mapeo de impuestos actualmente en la aeat es sobre impuestos, no se puede diferenciar entre el soportado y el repercutido, por lo que los cálculos de las casillas los unifican y no deberían (no es posible realizar una diferenciación por el debe o el haber debido que se necesitan ambos en ciertos casos).

Propuestas:
1.- Realizar el mapping de impuestos como lo realiza odoo mediante etiquetas.
2.- Realizar el mapping de impuestos como hasta ahora pero teniendo en cuenta las cuentas contables. 

En mi opinión el punto 1 de la propuesta me parece la mejor dado que es por donde Odoo plantea su solución, además está opción da más posibilidades a adaptar cuentas/subcuentas en casos de necesidad (ya sabéis los contables), a parte también da la opción de separación de localizaciones.
El problema que veo es realizar todos los mapeos, de una manera agil.

¿Qué pensáis al respecto?

Un saludo,

HaraldPanten commented:
También pienso que debería ser mediante etiquetas; lo veo mejor trabajado.

Aún así, entiendo que sería mejor crear nuevas "etiquetas OCA", no? Lo digo de cara a posibles cambios que pueda implementar Odoo o decisiones que no estén alineadas con l10n_es_aeat.

En caso de ser así, cuál creéis que sería la mejor forma de relacionar una etiqueta nueva con cada elemento del account.tax.repartition.line?

Respecto la opción 2, creo que quizá no sería del todo "fiable" si el criterio es basarse en las cuentas contables.

Se os ocurre alguna otra alternativa? @ValentinVinagre @pedrobaeza @MiquelRForgeFlow 

Un saludo!

pedrobaeza commented:
No he examinado el problema que estáis comentando, pero con etiquetas ya os digo que prefiero que no sea, porque se pierde la mejor característica del actual motor, y que precisamente es lo que provocaría muchas historias: el dinamismo de las definiciones de impuestos. Si lo basamos en etiquetas que se graban a fuego en el apunte y hay cualquier cambio en la legislación, no será retro-compatible (ni tampoco programable a futuro), y tocará hacer el cambio justo en el momento o tener que tocar apuntes a mano. ¿Me seguís?

ValentinVinagre commented:
Hola @pedrobaeza ,
Entiendo, pero entonces debería de poderse diferenciar por cuenta contable también, no solamente por impuesto, ya que si no, no se puede diferenciar entre los diferentes que ha unificado Odoo, podría configurarse por una cuenta padre (472%, 477%....) y que busque por las hijas también esa y por el impuesto. Si quieres esperamos a que lo reviséis un poco y lo hablamos.

pedrobaeza commented:
Sí, todavía no puedo emitir un juicio de valor correcto, pero tampoco quiero tener que crear subcuentas contables para cada porcentaje.

ValentinVinagre commented:
Disculpa creo que me he explicado mal. Lo que me refería en la configuración de la línea de los modelos de aeat además de indicarle el impuesto pertinente también se le debería de indicar las cuentas contables pertinentes, por ejemplo el:
IVA 4% adquisición intracomunitario. Bienes corrientes.
Tiene 3 reparticiones:
-base
-soportado
-repercutido
Entonces el motor actual diferencia correctamente base de tasas, el problema que tiene es que como odoo ha juntado impuestos, creo que si se mantiene la estructura actual del motor, lo mejor sería realizar otro campo de búsqueda referente a las cuentas contables (como se indica en el screenshot), el tema de utilizar el 472%, 477%, etc es para dar la libertad al usuario para crear subcuentas y que el motor las tenga en cuenta, no es que necesariamente de deban de crear.
No hay problema @pedrobaeza cuando podáis.
 
<img width="1402" alt="Captura de pantalla 2020-03-30 a las 12 07 36" src="https://user-images.githubusercontent.com/22913794/77901024-342d0480-727f-11ea-8ce9-9963ae619017.png">

HaraldPanten commented:
Como comenta @ValentinVinagre , está la opción de guiar el mapeo por las cuentas contables. Por las cabeceras al menos, que creo que no deberían cambiar.

Otra alternativa también, sería podernos basar directamente en las etiquetas que pone el l10n_es de Odoo, aunque desconozco si sería la mejor postura teniendo en cuenta lo que comenta @pedrobaeza . Además, no sé hasta que punto nos "salvaríamos" de los cambios de legislación y "retro-compatibilidades".

ValentinVinagre commented on l10n_es_aeat/models/l10n_es_aeat_map_tax_line.py:
```diff
@@ -0,0 +1,51 @@
+        default="yes",
+    )
+    inverse = fields.Boolean(string="Inverse summarize sign", default=False)
+    to_regularize = fields.Boolean(string="To regularize")
```
Se añade para poder configurar un campo string para filtrado de cuentas que admita wilcard con el simbolo '%'.

`

    account_expression = fields.Char(
        string="Account Expression",
        help="Only admin full account or account+%, split whit ','"
    )

    @api.constrains("account_expression")
    def _check_account_expression(self):
        for sel in self.filtered(lambda x: x.account_expression):
            for account in re.split(",[\s*]?", sel.account_expression):
                if not re.fullmatch(r"([0-9]+?%?)", account):
                    raise ValidationError(
                        _("The expression '%s' is not correct.") % sel.account_expression
                    )
`

ValentinVinagre commented on l10n_es_aeat/views/aeat_tax_code_mapping_view.xml:
```diff
@@ -0,0 +1,79 @@
+                <field name="name" />
+                <field name="to_regularize" />
+                <field name="tax_ids" widget="many2many_tags" />
+                <field name="move_type" />
```
mostrar después del tax_ids:
`<field name="account_expression" />`

ValentinVinagre commented on l10n_es_aeat/views/aeat_tax_code_mapping_view.xml:
```diff
@@ -0,0 +1,79 @@
+                    </group>
+                    <group>
+                        <field name="tax_ids" />
+                        <field name="move_type" />
```
mostrar después del tax_ids:
<field name="account_expression" />

ValentinVinagre commented on l10n_es_aeat/models/l10n_es_aeat_report_tax_mapping.py:
```diff
@@ -0,0 +1,194 @@
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+from odoo import _, api, exceptions, fields, models
+
```
Añadir
`from odoo.models import expression`

ValentinVinagre commented on l10n_es_aeat/models/l10n_es_aeat_report_tax_mapping.py:
```diff
@@ -0,0 +1,194 @@
+
+    def _get_partner_domain(self):
+        return []
+
```
Se añade la función para la creación del domain de cuentas

`


    def _account_codes_to_domain(self, account_codes):
        """Convert a comma separated list of account codes
        (possibly with wildcards) to a domain on account.account.
         """
        elems = []
        for account_code in account_codes.split(","):
            account_code = account_code.strip()
                if "%" in account_code:
                    elems.append([("account_id.code", "=like", account_code)])
                else:
                    elems.append([("account_id.code", "=", account_code)])
        return tuple(expression.OR(elems))
`

ValentinVinagre commented on l10n_es_aeat/models/l10n_es_aeat_report_tax_mapping.py:
```diff
@@ -0,0 +1,194 @@
+                ("tax_line_id", "in", taxes.ids),
+                ("tax_ids", "in", taxes.ids),
+            ]
+        if map_line.sum_type == "debit":
```
Se añade la comprobación para la creación del dominio de cuentas:
`

    if map_line.account_expression:
        move_line_domain += self._account_codes_to_domain(
            map_line.account_expression)`

ValentinVinagre commented on l10n_es_aeat/models/l10n_es_aeat_map_tax_line.py:
```diff
@@ -0,0 +1,51 @@
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl
+
+from odoo import fields, models
+
```
from odoo.exceptions import ValidationError
import re

ValentinVinagre reviewed (commented):
Hola @MiquelRForgeFlow ,
Si realizas los cambios indicados sería posible filtrar por cuentas contables según se necesite en las casillas de los informes.

He optado al final en una propuesta similar al miss reports, que da mas facilidad al realizar el filtraje (y también si Odoo cambia id de cuentas no se ven afectados los modelos) ¿que os parece @pedrobaeza @MiquelRForgeFlow  @HaraldPanten ?

HaraldPanten commented:
Creo que es una buena solución, ya que así cubrimos los casos (aunque sean pocos) de subcuentas de hacienda. Además, es un concepto que también se utiliza en las plantillas del MIS el hecho de mapear por cuentas además de padres con el %.

Si se toma esta decisión, sólo quedaría adaptar los modelos según esta lógica.

pedrobaeza commented:
El problema de esa opción es el rendimiento, ya que hacer un ilike lo destroza. Haré profiling a ver.

acysos commented on l10n_es_aeat/wizard/export_to_boe.py:
```diff
@@ -0,0 +1,245 @@
+            {
+                "name": file_name,
+                "datas": file,
+                "datas_fname": file_name,
```
El campo ya no se llama datas_fname se llama store_fname

MiquelRForgeFlow replied:
No tiene relacion con file_name. Simplemente el campo datas_fname se fusionó con el campo name. Gracias, se me había pasado.

ValentinVinagre commented on l10n_es_aeat/models/l10n_es_aeat_report_tax_mapping.py:
```diff
@@ -0,0 +1,194 @@
+                "|",
+                ("tax_line_id", "in", taxes.ids),
+                ("tax_ids", "in", taxes.ids),
+            ]
```
@MiquelRForgeFlow el nuevo cambio sería,
if map_line.campo_many2many_nuevo:
move_line_domain.append(("campo_cuenta", "in", mape_line.campo_many2many_nuevo.ids))

acysos commented:
@MiquelRForgeFlow Puedes hacer rebase squash con el anterior commit, y quede en uno. Filter account con Migration to 13.0.

pedrobaeza commented on .isort.cfg:
```diff
@@ -10,4 +10,4 @@ known_odoo=odoo
 sections=FUTURE,STDLIB,THIRDPARTY,ODOO,ODOO_ADDONS,FIRSTPARTY,LOCALFOLDER
 default_section=THIRDPARTY
-known_third_party=openerp,requests,setuptools
+known_third_party=requests,setuptools
```
esto no debería hacer falta tocarlo. De hecho, debería ir fuera hasta donde sé.

pedrobaeza replied:
Voy a poner más comentarios pero se me ha escapado lo de single comment.

MiquelRForgeFlow replied:
esto es automático por el script del isort, no es que lo quitara expresamente. antes aparecía el openerp porque había algun import que aun lo usaba. ahora todos los import son con odoo y por eso el isort quitó eso

pedrobaeza replied:
Ya, pero es que luego se "arregló" ese problema y con lo de `THIRDPARTY` es suficiente. Puedes eliminar la línea completa.

pedrobaeza commented on l10n_es_aeat/__manifest__.py:
```diff
@@ -10,17 +10,17 @@
     "category": "Accounting & Finance",
-    "depends": ["account", "base_iban", "account_tax_balance", "l10n_es"],
-    "external_dependencies": {"python": ["unidecode", "OpenSSL", "zeep", "requests"]},
+    "depends": ["l10n_es", "account_tax_balance"],
```
Por qué ahora depende de `l10n_es`?

MiquelRForgeFlow replied:
ya dependía antes, fíjate bien

acysos replied:
Al ver las dependencias externas que se quitan. Falta las del webservices soap y el certificado.

MiquelRForgeFlow replied:
?

pedrobaeza replied:
Cierto lo del `l10n_es`. El auto-formato me ha pasado una jugarreta. Las dependencias externas supongo que se referirá a `zeep`.

MiquelRForgeFlow replied:
zeep ya es requerimiento de odoo

pedrobaeza commented on l10n_es_aeat/models/l10n_es_aeat_report.py:
```diff
@@ -148,7 +141,7 @@ def _default_export_config_id(self):
         states={"draft": [("readonly", False)]},
     )
-    type = fields.Selection(
+    statement_type = fields.Selection(
```
Esto implica hacer un script de migración en cada declaración. No sé si merece la pena...

acysos replied:
En el resto de los módulos ya esta cambiado. El PR esta hecho teniendo en cuenta esto. Pero es correcto que obliga a cambiar la migración.

pedrobaeza replied:
Sí, y eso, el script de migración que incluya el `rename_fields`. Se puede hacer, pero que no se olvide.

MiquelRForgeFlow replied:
pues yo creo que si merece la pena, que es algo puntual y simple de añadir

pedrobaeza replied:
A ver, simple no es porque cada módulo `l10n_es_aeat_modX` necesita añadir el script de migración con `rename_fields`, pero que podemos llegar a hacerlo.

pedrobaeza commented on l10n_es_aeat/models/l10n_es_aeat_report.py:
```diff
@@ -311,50 +307,67 @@ def onchange_company_id(self):
+    @api.depends("year", "period_type")
+    def _compute_dates(self):
+        for report in self:
+            if not report.year or not report.period_type:
```
Puesto que es almacenado, yo no lo pondría a False, si no que los dejaría los que hubieran.

pedrobaeza commented on l10n_es_aeat/models/l10n_es_aeat_report.py:
```diff
@@ -311,50 +307,67 @@ def onchange_company_id(self):
+    @api.depends("year", "period_type", "date_start")
+    def _compute_export_config_id(self):
+        for report in self:
+            if not report.year or not report.period_type:
```
Reduce código aquí:
```
date = report.date_start or fields.Date.today()
report.export_config_id = report._get_export_config(date)
```

pedrobaeza commented on l10n_es_aeat/models/l10n_es_aeat_report.py:
```diff
@@ -311,50 +307,67 @@ def onchange_company_id(self):
+                        % (report.year, month, monthrange(report.year, month)[1])
+                    )
+
+    @api.depends("year", "period_type", "date_start")
```
No dependas de `year` ni de `period_type`

pedrobaeza commented on l10n_es_aeat/models/l10n_es_aeat_report.py:
```diff
@@ -492,12 +492,11 @@ def _prepare_aeat_sequence_vals(self, sequence, aeat_num, company):
-    @api.model_cr
     def _register_hook(self, companies=None):
         if not companies:
-            res = super(L10nEsAeatReport, self)._register_hook()
```
Siempre se debe tratar el valor de retorno como buena práctica aunque sea `None`, así que por favor no lo quites.

pedrobaeza commented on l10n_es_aeat/models/l10n_es_aeat_report_tax_mapping.py:
```diff
@@ -74,19 +68,18 @@ def _prepare_tax_line_vals(self, map_line):
+    def _get_move_line_domain(self, date_start, date_end, map_line):
         self.ensure_one()
         taxes = self.get_taxes_from_templates(map_line.tax_ids)
+        account = self.get_account_from_template(map_line.account_id)
```
Esto dentro del `if map_line.account_id:`

pedrobaeza commented on l10n_es_aeat/security/aeat_security.xml:
```diff
@@ -15,6 +15,6 @@
         <field
             name="domain_force"
-        >['|', ('company_id', '=', False), ('company_id', 'child_of', [user.company_id.id])]</field>
+        >['|', ('company_id', '=', False), ('company_id', 'in', company_ids)]</field>
```
Esto require script de migración:

https://github.com/OCA/maintainer-tools/wiki/Migration-to-version-13.0#tasks-to-do-in-the-migration

MiquelRForgeFlow replied:
eso si fuera noupdate=1, pero es noupdate=0, por lo que no creo que haga falta

pedrobaeza replied:
Vale, pues eso es un error a priori, porque debería ser noupdate="1". Cámbialo si no y script no hace falta, aunque todos los migrados seguirán con noupdate=0

pedrobaeza commented on l10n_es_aeat/static/src/js/aeat_report_client_action.js:
```diff
@@ -2,15 +2,11 @@ odoo.define("l10n_es_aeat.aeat_report_client_action", function(require) {
-    var QWeb = core.qweb;
 
-    var report_client_action = AbstractAction.extend(ControlPanelMixin, {
+    var report_client_action = AbstractAction.extend({
```
Puedes revisar los warning JS?

MiquelRForgeFlow replied:
¿Cuales? :S

pedrobaeza replied:
OK, ya veo que están quitadas.

pedrobaeza commented on l10n_es_aeat/tests/test_l10n_es_aeat_mod_base.py:
```diff
@@ -8,9 +8,8 @@
-@common.at_install(False)
-@common.post_install(True)
-class TestL10nEsAeatModBase(common.TransactionCase):
+@common.tagged("post_install", "-at_install")
```
¿Esto necesita seguir siendo post_install?

MiquelRForgeFlow replied:
ni idea

pedrobaeza replied:
Quítalo y lo comprobaremos en el futuro.

pedrobaeza reviewed (changes requested):
Bien en general y el nuevo campo está bien implementado, solo algunos comentarios generales de migración

ValentinVinagre commented on l10n_es_aeat/models/l10n_es_aeat_report_tax_mapping.py:
```diff
@@ -0,0 +1,197 @@
+        res = super().calculate()
+        for report in self:
+            report.tax_line_ids.unlink()
+            # Buscar configuración de mapeo de impuestos
```
Hola @MiquelRForgeFlow ,
Cuando se realiza el unlink de 'tax_line_ids', da un problema con los campos calculados, esto hace que se lancen los triggers con el campo 'tax_line_ids' vacío, para solucionar esto se puede obligar a Odoo a escribir en BBDD para que cuando cree las lineas se lancen correctamente los triggers otra vez.
```
report.flush()
report.invalidate_cache()
```

MiquelRForgeFlow replied:
gracias :+1:

pedrobaeza commented:
Pues vamos allá!

/ocabot merge nobump

pedrobaeza commented:
Hay un problema con las traducciones. Voy a arreglarlo.

pedrobaeza committed:
[FIX] l10n_es_aeat: Fix translations

pedrobaeza commented:
/ocabot merge nobump

pedrobaeza commented:
/ocabot merge nobump

pedrobaeza commented:
@sbidoul I think I have found the pattern for the merge commands that are not finished: if after the merge command, the bot emits the message of the approve label, it gets stuck on the merge. You can see in this timeline that sequence.
